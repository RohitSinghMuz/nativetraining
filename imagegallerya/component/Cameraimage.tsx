import React, {Component} from 'react';
import {
  StyleSheet,
  Text,
  View,
  PermissionsAndroid,
  TouchableOpacity,
  Image,
  SafeAreaView,
  FlatList,
  ScrollView,
  Platform,
} from 'react-native';
import ActionSheet from 'react-native-actionsheet';
import Ionicons from 'react-native-vector-icons/Ionicons';
import {launchCamera, launchImageLibrary} from 'react-native-image-picker';
interface IProps {}
interface IState {
  img: any;
  gallery: any;

  data: any[];
}

const optionarray = [
  'Select From File',
  'Captures with camera',
  'Close',
  // 'option4',
];

class Cameraimage extends Component<IProps, IState> {
  actionSheet: any;

  constructor(props: IProps) {
    super(props);
    this.state = {
      img: '',
      data: [],
      gallery: '',
    };
    this.actionSheet = React.createRef();
  }

  showActionSheet = () => {
    this.actionSheet.current.show();
  };

  handlePress = (index: any) => {
    // Alert.alert(optionarray[index]);
    if (index === 0) {
      this.imageGalleryLaunch();
    } else if (index === 1) {
      this.cameraLaunch();
    }
  };

  options: any = {
    saveToPhotos: true,
    mediaType: 'photo',
    allowsMultipleSelection: true,
  };

  imageGalleryLaunch = async () => {
    const result: any = await launchImageLibrary(this.options);
    console.log(result, 'result');
    result && result.assets && result.assets.length > 0 && result.assets[0].uri
      ? this.setState({gallery: result.assets[0].uri}, () => {
          this.setState({data: [...this.state.data, this.state.gallery]});
        })
      : null;
  };

  cameraLaunch = async () => {
    const granted: any = await PermissionsAndroid.request(
      PermissionsAndroid.PERMISSIONS.CAMERA,
    );
    console.log(granted, PermissionsAndroid.RESULTS.GRANTED, 'granted');

    if (granted === PermissionsAndroid.RESULTS.GRANTED) {
      const CameraResult: any = await launchCamera(this.options);
      console.log(CameraResult, 'CameraResult');
      CameraResult &&
      CameraResult.assets &&
      CameraResult.assets.length > 0 &&
      CameraResult.assets[0].uri
        ? this.setState({gallery: CameraResult.assets[0].uri}, () => {
            this.setState({data: [...this.state.data, this.state.gallery]});
          })
        : null;
    }
  };

  deletetext = (idx: any) => {
    const deleteitem = this.state.data?.filter((item: any, i: any) => i != idx);
    this.setState({
      data: deleteitem,
    });
  };

  render() {
    console.log(this.state.data, 'data');
    return (
      <SafeAreaView
        style={{
          backgroundColor: '#75E6DA',
          flex: 1,
        }}>
        <ScrollView>
          <View>
            <View style={styles.centeredView}>
              <FlatList
                testID="flatlistitem"
                // numColumns={2}
                contentContainerStyle={{
                  width: 400,
                  height: 300,
                }}
                data={this.state.data}
                keyExtractor={(index: number) => index.toString()}
                renderItem={({item, index}: any) => {
                  return (
                    <View style={styles.Imagedisplay}>
                      <Image
                        key={index}
                        testID="ImagedisplayID"
                        style={{
                          height: 150,
                          width: 150,
                          borderRadius: 10,
                          margin: 10,
                        }}
                        source={{
                          uri: item,
                        }}
                      />
                      <View>
                        <Ionicons
                          testID="deleteImage"
                          onPress={() => this.deletetext(index)}
                          name="ios-close-sharp"
                          size={25}
                          color="red"
                          style={{marginLeft: 70, fontSize: 30}}
                        />
                      </View>
                    </View>
                  );
                }}
              />
            </View>

            <View style={styles.actionbtnfile}>
              <TouchableOpacity
                testID="actionBtnText"
                onPress={this.showActionSheet}>
                <Text style={styles.actionbtn}>Open bottom action</Text>
              </TouchableOpacity>
            </View>
            <View testID="actionSheet">
              <ActionSheet
                ref={this.actionSheet}
                // title={'actionsheetexample'}
                options={optionarray}
                cancelButtonIndex={2}
                onPress={this.handlePress}
              />
            </View>
          </View>
        </ScrollView>
      </SafeAreaView>
    );
  }
}

export default Cameraimage;
const styles = StyleSheet.create({
  actionbtnfile: {
    marginTop: 300,
    flex: 1,
    justifyContent: 'center',
    alignContent: 'center',
    alignItems: 'center',
    position: 'relative',
  },
  actionbtn: {
    justifyContent: 'center',
    alignSelf: 'center',
    alignItems: 'center',
    backgroundColor: 'black',
    color: 'white',
    fontWeight: '700',
    padding: 10,
    borderRadius: 8,
    fontSize: 25,
  },
  Imagedisplay: {
    flex: 1,
    // flexDirection: 'row',

    justifyContent: 'flex-start',
    margin: 10,
  },
  centeredView: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: '#75E6DA',
    alignItems: 'center',
  },
  centeredViewModal: {
    flex: 1,
    justifyContent: 'center',
    backgroundColor: 'rgba(0,0,0,.6)',
    alignItems: 'center',
  },
  modalView: {
    width: '100%',
    backgroundColor: '#AAE3E2',
    borderRadius: 20,
    padding: 35,
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
    position: 'absolute',
    bottom: 0,
    left: 0,
  },
  button: {
    borderRadius: 20,
    padding: 10,
    elevation: 2,
  },
  buttonOpen: {
    marginVertical: 100,
    backgroundColor: '#A0E7E5',
    height: 60,
  },
  buttonClose: {
    backgroundColor: '#189AB4',
    margin: 20,
  },
  buttonClosen: {
    backgroundColor: '#2196F3',
  },
  textStyle: {
    color: '#0C2D48',
    fontWeight: 'bold',
    textAlign: 'center',
    padding: 5,
    fontSize: 22,
  },
  textStyleClose: {
    color: '#0C2D48',
    fontWeight: 'bold',
    textAlign: 'center',
    paddingVertical: 10,
    paddingHorizontal: 50,
    fontSize: 22,
  },
  modalText: {
    marginBottom: 15,
    textAlign: 'center',
  },
});
