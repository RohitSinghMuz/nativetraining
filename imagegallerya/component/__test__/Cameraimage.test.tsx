import React from 'react';
import {PermissionsAndroid, View as MockView} from 'react-native';
import {render, screen, fireEvent, act} from '@testing-library/react-native';
import Cameraimage from '../Cameraimage';
import Ionicons from 'react-native-vector-icons/Ionicons';
import ActionSheet from 'react-native-actionsheet';
import {launchCamera} from 'react-native-image-picker';

jest.mock('react-native-image-picker', () => ({
  launchCamera: jest.fn().mockImplementation(() => ({
    assets: [
      {
        uri: 'filepath',
      },
    ],
  })),
  launchImageLibrary: jest.fn().mockImplementation(() => ({
    assets: [
      {
        uri: 'filepath',
      },
    ],
  })),
}));

jest.mock('react-native/Libraries/Utilities/Platform', () => ({
  OS: 'android',
  select: () => null,
}));

jest.mock('react-native-vector-icons/Ionicons', () => (props: any) => (
  <MockView {...props} />
));
class mockClass extends React.Component {
  render(): React.ReactNode {
    return <></>;
  }
}
jest.mock('react-native-actionsheet', () => {
  return (props: any) => {
    console.log('-----clg', props);
    props.onPress();
    return <></>;
  };
});

jest.spyOn(React, 'createRef').mockImplementation(() => ({
  current: {
    show: jest.fn(),
  },
}));

describe('Test cases of  Cameraimage component', () => {
  it('onPress function is called when "Select File" button is pressed', async () => {
    const {getByTestId} = render(<Cameraimage />);
    const selectFileButton: any = getByTestId('actionBtnText');
    // console.log('selectFileButton', selectFileButton);
    fireEvent.press(selectFileButton);
    // expect(mockFunction).toHaveBeenCalled();
  });

  it('requests for PermissionsAndroid camera', async () => {
    const permission = PermissionsAndroid.PERMISSIONS.CAMERA;
    const granted: any = true;
    console.log(permission, 'permission');
    jest
      .spyOn(PermissionsAndroid, 'request')
      .mockImplementation(() => Promise.resolve('granted'));

    await expect(PermissionsAndroid.request(permission)).resolves.toBe(
      'granted',
    );
    expect(PermissionsAndroid.request).toHaveBeenCalledWith(permission);
  });

  it('imageGalleryLaunch tst cases', async () => {
    const {getByTestId} = render(<Cameraimage />);
    const actionSheet: any = getByTestId('actionSheet');
    // console.log(actionSheet.props.children.props, 'actionSheet');
    actionSheet.props.children.props.onPress(0);
  });

  it('cameraLaunch tst cases', async () => {
    const {getByTestId} = render(<Cameraimage />);
    const actionSheet: any = getByTestId('actionSheet');
    actionSheet.props.children.props.onPress(1);
  });

  it('should set the state with the gallery uri if result contains assets', async () => {
    const assets: any = [
      {
        uri: 'filepath',
      },
    ];

    jest.fn().mockImplementation(assets);
  });

  // it('display flatlist item', () => {
  //   const {getByTestId} = render(<Cameraimage />);
  //   const flatlist: any = getByTestId('flatlistitem');
  //   expect(getByTestId('flatlist')).toBeDefined();
  // });

  it('calls deletetext function when button is pressed', () => {
    const {getByTestId} = render(<Cameraimage />);
    const flatlist: any = getByTestId('flatlistitem');
    const renderItemWrapper = render(
      flatlist.props.renderItem({item: 'testur.in', index: 0}),
    );
    const deletBtn = renderItemWrapper.getByTestId('deleteImage');
    console.log('deletBtn', deletBtn);
    fireEvent.press(deletBtn);
  });
});
