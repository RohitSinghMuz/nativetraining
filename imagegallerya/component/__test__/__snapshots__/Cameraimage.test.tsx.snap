// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test cases of  Cameraimage component calls deletetext function when button is pressed 1`] = `
<View
  style={
    {
      "backgroundColor": "#75E6DA",
      "flex": 1,
    }
  }
>
  <RCTScrollView>
    <View>
      <View>
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#75E6DA",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <RCTScrollView
            contentContainerStyle={
              {
                "height": 300,
                "width": 400,
              }
            }
            data={[]}
            getItem={[Function]}
            getItemCount={[Function]}
            keyExtractor={[Function]}
            onContentSizeChange={[Function]}
            onLayout={[Function]}
            onMomentumScrollBegin={[Function]}
            onMomentumScrollEnd={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            removeClippedSubviews={true}
            renderItem={[Function]}
            scrollEventThrottle={50}
            stickyHeaderIndices={[]}
            testID="flatlistitem"
            viewabilityConfigCallbackPairs={[]}
          >
            <View />
          </RCTScrollView>
        </View>
        <View
          style={
            {
              "alignContent": "center",
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "marginTop": 300,
              "position": "relative",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="actionBtnText"
          >
            <Text
              style={
                {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "backgroundColor": "black",
                  "borderRadius": 8,
                  "color": "white",
                  "fontSize": 25,
                  "fontWeight": "700",
                  "justifyContent": "center",
                  "padding": 10,
                }
              }
            >
              Open bottom action
            </Text>
          </View>
        </View>
        <View
          testID="actionSheet"
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
